In [1]:
# Initialize task.
import jax.numpy as jnp
import numpy as np
import matplotlib.pyplot as plt
import seaborn as snb
import matplotlib.colors as colors
from scipy.stats import norm
from scipy.optimize import minimize
from jax import hessian

snb.set_theme(font_scale=1.25)

# Load Data

data = jnp.load('./data_exercise5b.npz')
X = data['day']
y = np.log(data['bike_count'])

#Standardize data
ym, ys = jnp.mean(y), jnp.std(y)
y = (y - ym) / ys

$\textbf{Part 1: Fully Bayesian inference for Gaussian process regression}$

$\textbf{Task 1.1}$ Choose a value for $\textit{v}$ such that the prior probability of observing a lengthscale larger than 100 is approximately $1\%$

We identify firs that our lengthscale parameter follows af half normal distribution given in 

\begin{equation}
    \mathcal{l} \sim \mathcal{N}_+(0,v)
\end{equation}

We now the condition we need to fulfill as

\begin{equation}
    P(\mathcal{l} > 100) \approx 0.01
\end{equation}

We can express our distribution of $l$ as $l = |\sqrt{v} \cdot Z|$ with $Z$ being a standard normal variable. This is because we now that $l$ is a half normal distributions. We can then insert this equation into our condition. We do some reductions and since we have that the standard normal distribution is symmetric we can view it as a two-tailed probability.

\begin{align}
P(l > 100) = P(|\sqrt{v} \cdot Z| > 100) &\approx 0.01 \\
P(\frac{|Z| > 100}{\sqrt{v}}) &\approx 0.01 \hspace{2cm} \text{Two-tailed} \\
2 \cdot P(\frac{Z > 100}{ \sqrt{v}}) &\approx 0.01 \\
P(\frac{Z > 100}{ \sqrt{v}}) &\approx 0.005

\end{align}

We now use the normal CDF value of z to find the z_value that satisfies this. We can either look up a table our use python.

In [12]:
z_value = norm.ppf(1-0.005)
print(f"z_value: {z_value}")

z_value: 2.5758293035489004


We can now solve the expression for v

\begin{align}
    \frac{100}{\sqrt{v}} = z_{value} \\
    v = \left(\frac{100}{z_{value}}\right) ^2
\end{align}

In [14]:
v = (100 / z_value) ** 2
print(f"v: {v}")

v: 1507.1824930113974


So when $v \approx 1507$ the probability of the lengthscale parameter being above 100 will be roughly 1%